<a href="https://colab.research.google.com/github/AnoopGeorge418/MACHINE-LEARNING-PROJECTS/blob/main/Rock%20vs%20Mine%20Prediction/SONAR-Rock-vs-Mine-Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Problem Defination - To create a model that helps the submarines to predict a mine or rock under the sea during the war.**

# **2. Work Flow: Supervised learning - classification - Logestic regression**
- **Collect Sonar data**
- **Data PreProcessing**
- **Train Test Split the data**
- **Load to ML model**
- **Get Trained model**
- **Model Prediction**
- **Creating a predictive system**

In [1]:
# importing dependancies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
%matplotlib inline

# **Data Collection**

In [2]:
sonar_data = pd.read_csv('/content/sonar data.csv', header = None)
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

# **Data Preprocessing**

In [3]:
sonar_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

In [4]:
sonar_data.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [5]:
# checking the counts of rocks and mines
sonar_data[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

In [6]:
# grouping 60 and getting mean - there is a diffrence in mean of both r and m
sonar_data.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [7]:
# separating columns
X = sonar_data.iloc[:, :-1] # Independent Column
Y = sonar_data.iloc[:, -1] # Target/Dependent column

In [8]:
print(X, Y)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

# **Splitting data**

In [9]:
# Splitting for Training and Testing data
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.1, stratify = Y, random_state = 1)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(187, 60) (21, 60) (187,) (21,)


# **Model training**

In [10]:
# Model training - Logestic Regression
model = LogisticRegression()
model

LogisticRegression()

In [11]:
# Training model
model.fit(X_train, y_train)

LogisticRegression()

# **Model Evaluation**

In [12]:
# Accuracy score on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, y_train)
print(f'Accuracy Score: {training_data_accuracy}')

Accuracy Score: 0.8342245989304813


In [13]:
# Accuracy score on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, y_test)
print(f'Accuracy Score: {test_data_accuracy}')

Accuracy Score: 0.7619047619047619


# **Creating a predictive system** - checking how well model works on  unseen data

In [16]:
# new data
input_data = (0.0286,0.0453,0.0277,0.0174,0.0384,0.0990,0.1201,0.1833,0.2105,0.3039,0.2988,0.4250,0.6343,0.8198,1.0000,0.9988,0.9508,0.9025,0.7234,0.5122,0.2074,0.3985,0.5890,0.2872,0.2043,0.5782,0.5389,0.3750,0.3411,0.5067,0.5580,0.4778,0.3299,0.2198,0.1407,0.2856,0.3807,0.4158,0.4054,0.3296,0.2707,0.2650,0.0723,0.1238,0.1192,0.1089,0.0623,0.0494,0.0264,0.0081,0.0104,0.0045,0.0014,0.0038,0.0013,0.0089,0.0057,0.0027,0.0051,0.0062)

# converting input_data to numpy array
input_data_as_array = np.asarray(input_data)

# reshaping data cause we are predicting for 1 instance
reshaped_data = input_data_as_array.reshape(1, -1)

# making prediction
prediction = model.predict(reshaped_data)
if prediction == 'R':
    print(f'Predicted Value: "Rock", for New data: {input_data}')
else:
    print(f'Predicted value: "Mine", for New data: {input_data}')

Predicted Value: "Rock", for New data: (0.0286, 0.0453, 0.0277, 0.0174, 0.0384, 0.099, 0.1201, 0.1833, 0.2105, 0.3039, 0.2988, 0.425, 0.6343, 0.8198, 1.0, 0.9988, 0.9508, 0.9025, 0.7234, 0.5122, 0.2074, 0.3985, 0.589, 0.2872, 0.2043, 0.5782, 0.5389, 0.375, 0.3411, 0.5067, 0.558, 0.4778, 0.3299, 0.2198, 0.1407, 0.2856, 0.3807, 0.4158, 0.4054, 0.3296, 0.2707, 0.265, 0.0723, 0.1238, 0.1192, 0.1089, 0.0623, 0.0494, 0.0264, 0.0081, 0.0104, 0.0045, 0.0014, 0.0038, 0.0013, 0.0089, 0.0057, 0.0027, 0.0051, 0.0062)


In [19]:
# new data
input_data = (0.0530,0.0885,0.1997,0.2604,0.3225,0.2247,0.0617,0.2287,0.0950,0.0740,0.1610,0.2226,0.2703,0.3365,0.4266,0.4144,0.5655,0.6921,0.8547,0.9234,0.9171,1.0000,0.9532,0.9101,0.8337,0.7053,0.6534,0.4483,0.2460,0.2020,0.1446,0.0994,0.1510,0.2392,0.4434,0.5023,0.4441,0.4571,0.3927,0.2900,0.3408,0.4990,0.3632,0.1387,0.1800,0.1299,0.0523,0.0817,0.0469,0.0114,0.0299,0.0244,0.0199,0.0257,0.0082,0.0151,0.0171,0.0146,0.0134,0.0056)

# converting input_data to numpy array
input_data_as_array = np.asarray(input_data)

# reshaping data cause we are predicting for 1 instance
reshaped_data = input_data_as_array.reshape(1, -1)

# making prediction
prediction = model.predict(reshaped_data)
if prediction == 'R':
    print(f'Predicted Value: "Rock", for New data: {input_data}')
else:
    print(f'Predicted value: "Mine", for New data: {input_data}')

Predicted value: "Mine", for New data: (0.053, 0.0885, 0.1997, 0.2604, 0.3225, 0.2247, 0.0617, 0.2287, 0.095, 0.074, 0.161, 0.2226, 0.2703, 0.3365, 0.4266, 0.4144, 0.5655, 0.6921, 0.8547, 0.9234, 0.9171, 1.0, 0.9532, 0.9101, 0.8337, 0.7053, 0.6534, 0.4483, 0.246, 0.202, 0.1446, 0.0994, 0.151, 0.2392, 0.4434, 0.5023, 0.4441, 0.4571, 0.3927, 0.29, 0.3408, 0.499, 0.3632, 0.1387, 0.18, 0.1299, 0.0523, 0.0817, 0.0469, 0.0114, 0.0299, 0.0244, 0.0199, 0.0257, 0.0082, 0.0151, 0.0171, 0.0146, 0.0134, 0.0056)


# **Overall Working and Inference**


---


- # **Imported all necessary libaraies**
- # **loaded the data**
- # **Performed data preprocessing**
- # **Checked how many R and M in outcome and found out almost equal.**
- # **Grouped the data based on target variable and performed mean found there is a diffrence in mean values of R and M**
- # **Separated independent and dependent features**
- # **Splitted data for training and testing**
- # **loaded the training data to the model and fitted it**
- # **Performed model evaluation done on testing data using accuracy score and got**
    - # **83% on training data**
    - # **76% on testing data**
- # **Created a predictive system in order to check hw well the model performed on unseen/new data that passed.**


---



---

